plugins/package-managers/node/src/main/kotlin/Utils.kt

```diff
@@ -245,7 +246,37 @@ internal val PackageJson.moduleId: String get() =
 /**
  * Return the directories of all modules which have been installed in the 'node_modules' dir within [moduleDir].
  */
-internal fun getInstalledModulesDirs(moduleDir: File): Set<File> =
-    moduleDir.resolve("node_modules").walk().filter {
-        it.isFile && it.name == NodePackageManagerType.DEFINITION_FILE
-    }.mapTo(mutableSetOf()) { it.parentFile.realFile }
+internal fun getInstalledModulesDirs(projectDir: File): Set<File> {
+    val modulesToProcess = LinkedList<File>().apply { add(projectDir.realFile) }
+    val discoveredModules = mutableSetOf<File>()
+
+    while (modulesToProcess.isNotEmpty()) {
+        val currentModule = modulesToProcess.removeFirst()
+        if (!discoveredModules.add(currentModule)) continue
+
+        val childModules = getChildModuleDirs(currentModule)
+        modulesToProcess += childModules
+    }
+
+    return discoveredModules
+}
+
+/**
+* Find all direct dependency module directories within the node_modules directory of the given [moduleDir].
+* This handles both regular modules and namespaced (@organization) modules.
+*/
+private fun getChildModuleDirs(moduleDir: File): Set<File> {
+    val nodeModulesDir = moduleDir.resolve("node_modules").takeIf { it.isDirectory } ?: return emptySet()
+
+    fun File.isModuleDir(): Boolean =
+        isDirectory && !isHidden && !name.startsWith("@") && resolve(NodePackageManagerType.DEFINITION_FILE).isFile
+
+    val childModuleDirsWithoutNamespace = nodeModulesDir.listFiles().filter { it.isModuleDir() }
+    val childModuleDirsWithNamespace = nodeModulesDir.listFiles().filter {
```

Comment:
Prefer `walk().maxDepth(1)` over `listFiles()`. Also, probably exact the list to a variable, so the filesystem is only accessed once.